### PR TITLE
Allow to omit indices in index range expansions

### PR DESCRIPTION
**Pull request by krobelus**
## Description

So omitted range limits in, say `$PATH[..]` default to the first/last element, just like Python/Go/Rust slices.
This is something I have had expected to work several times (coming from Python). It essentially saves typing the `1`/`-1` in a `..` range and I don't see any harm in defaulting to those values.
However, it's not super well thought out, and a language change so I'm asking if there are any concerns.

## TODOs:
- [x] Changes to fish usage are reflected in user documentation/manpages.
- [x] Tests have been added for regressions fixed
- [x] User-visible changes noted in CHANGELOG.md

### CODE REVIEW

File: src/expand.cpp [hunk 1]
```diff
@@ -189,9 +189,14 @@ static size_t parse_slice(const wchar_t *in, wchar_t **end_ptr, std::vector<long
 
         const wchar_t *end;
         long tmp = fish_wcstol(&in[pos], &end);
-        // We don't test `*end` as is typically done because we expect it to not be the null char.
-        // Ignore the case of errno==-1 because it means the end char wasn't the null char.
-        if (errno > 0) {
+        if (idx.empty() && in[pos] == L'.' && in[pos + 1] == L'.') {
```
Comment (ridiculousfish) [inline src/expand.cpp]: This is confusing to read - we call `wcstol` but sometimes ignore the result. Can we rearrange the test so that we check for a missing index before calling `wcstol()`?

File: src/expand.cpp [hunk 2]
```diff
@@ -205,7 +210,16 @@ static size_t parse_slice(const wchar_t *in, wchar_t **end_ptr, std::vector<long
             long tmp1 = fish_wcstol(&in[pos], &end);
             // Ignore the case of errno==-1 because it means the end char wasn't the null char.
             if (errno > 0) {
-                return pos;
+                // Check if we are at the last index expression, a missing end index means the range
```
Comment (ridiculousfish) [inline src/expand.cpp]: I'd like to see this check hoisted as well. If the next char is a ']', then the index is omitted; otherwise call wcstol()

File: tests/test8.in [hunk 3]
```diff
@@ -4,11 +4,18 @@ logmsg Test variable expand
...
+echo $test[8.. ]       # range up to implied $n
 echo $test[$n..1]      # inverted range
 echo $test[2..5 8..6]  # several ranges
 echo $test[-1..-2]     # range with negative limits
 echo $test[-1..1]      # range with mixed limits
 
+# missing start, cannot use implied range
+echo $test[1..2 ..]
+echo $test[..1 ..2]
+
```
Comment (ridiculousfish) [inline tests/test8.in]: Please add a test that makes the behavior clear for an empty variable, e.g.:

    echo $test[..$empty]

I do not have a strong opinion about what that should do, but it should be something we decide and test.

File: src/expand.cpp (hunk 1 above)
Comment (krobelus) [reply src/expand.cpp]: Yep, done!

File: tests/test8.in (hunk 3 above)
Comment (krobelus) [reply tests/test8.in]: Ah right, I added a test for it.
Turns out this expands to no indices because of cartesian product expansion. It is at least consistent with the same code with 1 on the left side.

**Comment by faho** [comment]
So currently having an empty variable in a range is *allowed*, because it triggers the cartesian product (and hence removes the entire token). I.e. this doesn't error:

```fish
set -l empty
set -l list 1 2 3
echo $list[1..$empty]

# even this is allowed:
echo $list[..$empty]
```

Which means if we handled empty variables any differently here it would be a breaking change. I think the cartesian product here is consistent with it elsewhere and therefore the right thing, but then we've never been in love with it in general. However that would be a larger discussion, and I think it's one we could have later.

So I would keep this as-is, except for one test I'd like to add:

```fish
echo $list[.."$empty"]
echo $list["$empty"..]
```

This prints the entire list, which seems logical however you slice it.

Otherwise:

LGTM!

**Comment by krobelus** [comment]
> This prints the entire list, which seems logical however you slice it.

A+ pun right there :joy: :joy:

Let's ship it!
